utilities/table_properties_collectors/compact_for_tiering_collector.cc

```diff
@@ -0,0 +1,177 @@
+#include "rocksdb/types.h"
+#include "rocksdb/utilities/customizable_util.h"
+#include "rocksdb/utilities/object_registry.h"
+#include "rocksdb/utilities/options_type.h"
+#include "rocksdb/utilities/table_properties_collectors.h"
+#include "util/string_util.h"
+
+namespace ROCKSDB_NAMESPACE {
+const std::string
+    CompactForTieringCollector::kNumEligibleLastLevelEntriesPropertyName =
+        "rocksdb.eligible.last.level.entries";
+
+CompactForTieringCollector::CompactForTieringCollector(
+    int level_at_creation, int num_levels,
+    SequenceNumber last_level_inclusive_max_seqno_threshold,
+    size_t compaction_trigger, bool enabled)
+    : level_at_creation_(level_at_creation),
+      num_levels_(num_levels),
+      last_level_inclusive_max_seqno_threshold_(
+          last_level_inclusive_max_seqno_threshold),
+      compaction_trigger_(compaction_trigger),
+      enabled_(enabled) {}
+
+Status CompactForTieringCollector::AddUserKey(const Slice& /*key*/,
+                                              const Slice& value,
+                                              EntryType type,
+                                              SequenceNumber seq,
+                                              uint64_t /*file_size*/) {
+  if (!enabled_ || level_at_creation_ == num_levels_ - 1 ||
+      last_level_inclusive_max_seqno_threshold_ == kMaxSequenceNumber) {
+    return Status::OK();
+  }
+  SequenceNumber seq_for_check = seq;
+  if (type == kEntryTimedPut) {
+    seq_for_check = ParsePackedValueForSeqno(value);
+  }
+  if (seq_for_check < last_level_inclusive_max_seqno_threshold_) {
+    last_level_eligible_entries_counter_++;
+  }
+  return Status::OK();
+}
+
+Status CompactForTieringCollector::Finish(UserCollectedProperties* properties) {
+  assert(!finish_called_);
+  if (compaction_trigger_ != 0 &&
+      last_level_eligible_entries_counter_ >= compaction_trigger_) {
+    need_compaction_ = true;
+  }
+  finish_called_ = true;
+  if (last_level_eligible_entries_counter_ > 0) {
+    *properties = UserCollectedProperties{
+        {kNumEligibleLastLevelEntriesPropertyName,
+         std::to_string(last_level_eligible_entries_counter_)},
+    };
+  }
+  return Status::OK();
+}
+
+UserCollectedProperties CompactForTieringCollector::GetReadableProperties()
+    const {
+  return UserCollectedProperties{
+      {kNumEligibleLastLevelEntriesPropertyName,
+       std::to_string(last_level_eligible_entries_counter_)},
+  };
+}
+
+bool CompactForTieringCollector::NeedCompact() const {
+  return need_compaction_;
+}
+
+void CompactForTieringCollector::Reset() {
+  last_level_eligible_entries_counter_ = 0;
+  finish_called_ = false;
+  need_compaction_ = false;
+}
+
+TablePropertiesCollector*
+CompactForTieringCollectorFactory::CreateTablePropertiesCollector(
+    TablePropertiesCollectorFactory::Context context) {
+  uint64_t compaction_trigger = GetCompactionTrigger(context.column_family_id);
+  return new CompactForTieringCollector(
```

Comment:
This function can return nullptr to decline collecting properties when it's not going to do anything anyway (!enabled). That's more efficient than virtual calling into the collector and checking enabled_ on every user key of every compaction.